### PR TITLE
Deprecate `brew cask update`.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/master/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/homebrew/pull/49031).
- [x] Have you successfully run `brew tests` with your changes locally?

-----

In addition to deprecating `brew cask update` I also added a new option to `odeprecated` which allows us to specify a date after which a given method will be disabled.

---

Closes https://github.com/caskroom/homebrew-cask/issues/26921.

### CODE REVIEW

**Comment by vitorgalvao** [comment]
Also needs to be removed from the man page and `brew cask help`.

File: Library/Homebrew/compat/hbc/cli/update.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+module Hbc
```
Comment (MikeMcQuaid) [inline Library/Homebrew/compat/hbc/cli/update.rb]: I'd maybe go `Library/Homebrew/hbc/compat/cli/update.rb` for this.

File: Library/Homebrew/utils.rb [hunk 2]
```diff
@@ -44,24 +45,32 @@ def odie(error)
...
+def odeprecated(method, replacement = nil, disable: false, disable_on: nil, caller: send(:caller))
   replacement_message = if replacement
     "Use #{replacement} instead."
   else
     "There is no replacement."
   end
 
+  unless disable_on.nil?
+    if disable_on > Time.now
+      will_be_disabled_message = " and will be disabled on #{disable_on.strftime("%B %-d, %Y")}"
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Let's go for a "29 May 2017" date format.

File: Library/Homebrew/utils.rb [hunk 3]
```diff
@@ -44,24 +45,32 @@ def odie(error)
...
+    else
+      disable = true
+    end
+  end
+
+  verb = if disable
+    "disabled"
+  else
+    "deprecated".concat(will_be_disabled_message.to_s)
```
Comment (MikeMcQuaid) [inline Library/Homebrew/utils.rb]: Just interpolate this I think.

File: Library/Homebrew/utils.rb (hunk 2 above)
Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: Went with ISO 8601.

File: Library/Homebrew/utils.rb (hunk 3 above)
Comment (reitermarkus) [inline Library/Homebrew/utils.rb]: Done.

File: Library/Homebrew/compat/hbc/cli/update.rb (hunk 1 above)
Comment (reitermarkus) [inline Library/Homebrew/compat/hbc/cli/update.rb]: I think it's best to keep all of these in one central location.

Comment (MikeMcQuaid) [inline Library/Homebrew/compat/hbc/cli/update.rb]: Until the Homebrew Cask code is structured and the compatibility layer is a lot closer to the rest of Homebrew I disagree. Compared to the other compat modules this one is included as-is regardless of a `--no-compat` or similar, isn't included into Homebrew's `compat`.

Comment (reitermarkus) [inline Library/Homebrew/compat/hbc/cli/update.rb]: > this one is included as-is regardless of a `--no-compat`

No, it isn't. Everything under `compat/hbc` is not included when `--no-compat` is specified.

Comment (reitermarkus) [inline Library/Homebrew/compat/hbc/cli/update.rb]: Ah, right, I missed the `require "compat/hbc/cli/update"` line.

Comment (MikeMcQuaid) [inline Library/Homebrew/compat/hbc/cli/update.rb]: Ok, I see. I'm fine with this then 👍

**Comment by reitermarkus** [comment]
Just noticed that we do in fact have an easy way to hide the `update` command from `help`.

**Pull request merged**